### ISSUE TITLE
[BUG] Column-selected lines don't get trailing space removed when 'Trim Trailing Space' is commanded

### ISSUE THREAD

**Issue by alankilborn** [notepad-plus-plus/notepad-plus-plus#15743]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description of the Issue

See Steps To Reproduce (STR).

### Steps To Reproduce

1.  Create a new tab and make some lines with trailing spaces, e.g.
![image](https://github.com/user-attachments/assets/644aafe8-31b8-4c38-b06d-98ab6c2811ee)
2.  Make a column selection, e.g.
![image](https://github.com/user-attachments/assets/c2cbf9a9-9867-4be2-b284-2814347c722a)
3.  Execute _Edit_ > _Blank Operations_ > _Trim Trailing Space_ 

### Current Behavior

Nothing changes, i.e., trailing space not removed from selected lines.

### Expected Behavior

Trailing spaces on the selected lines should be removed.

### Debug Information

```shell
Notepad++ v8.7   (64-bit)
Build time : Sep 17 2024 - 17:06:31
Path : C:\NotepadPlusPlus\Misc\_NoBackup\Releases\npp.8.7\npp.8.7.portable.x64\notepad++.exe
Command Line : -titleAdd="8.7 64-bit" -multiInst
Admin mode : OFF
Local Conf mode : ON
Cloud Config : OFF
Periodic Backup : ON
OS Name : Windows 11 Pro (64-bit)
OS Version : 23H2
OS Build : 22631.4317
Current ANSI codepage : 1252
Plugins : 
    mimeTools (3.1)
    NppConverter (4.6)
    NppExport (0.4)
    PythonScript (2)
```

### Anything else?

Note that for a stream selection such as the following, 
![image](https://github.com/user-attachments/assets/473e08c4-e68d-4350-9201-5387a80968f5)

the trailing space _will_ be removed after the execution of the command:
![image](https://github.com/user-attachments/assets/75bb17eb-2b86-40b0-b895-803848c5f86d)

**Comment by Coises** [comment]
Just an additional data point: after a quick test, it appears that when there is a multiple selection, trailing spaces are removed only from the line(s) included in the primary selection.

For what it’s worth, personally, I don’t see why this matters much. I would have expected that when there is a selection, only trailing blanks included in the selection which are in lines in which the end of line character (or the last character in the file) is also included in the selection would be removed. So it’s a bit of a surprise to me that any line touched by the selection (if it a stream selection) is fully trimmed; but if that is the behavior, I suppose it makes sense to be consistent for column selections (and multiple selections?).

**Comment by Coises** [comment]
For what it’s worth, we can see here:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/682570ea6d364dd5d72453882e87d067148bb6be/PowerEditor/src/Notepad_plus.cpp#L1669
that column selections were intentionally excluded. So, technically, a request, not a bug.

**Comment by alankilborn** [comment]
@Coises

> it appears that when there is a multiple selection, trailing spaces are removed only from the line(s) included in the primary selection.

This is true; I tend to think of column selections _as_ multiple selections (which, technically, they are).
Technically, this issue came out of something I promised to @pryrt that I'd make an issue out of.  I suppose I was "checking off a todo item" rather than trying to be all-encompassing.
But sure, let's extend it to multiple stream selections as well.  I'll edit the title.

>personally, I don’t see why this matters much.

Well, some things we don't see matter a whole big bunch; see the somewhat related: https://github.com/editorconfig/editorconfig-notepad-plus-plus/issues/57

>expected that when there is a selection, only trailing blanks included in the selection which are in lines in which the end of line character...is also included in the selection would be removed

That might be a bit too much to expect.  The "lines" commands in Notepad++, when they work on a selection, do so on lines or partial lines selected.  I consistently call this operating on "lines touched by a selection" (as did you).

>column selections were intentionally excluded. So, technically, a request, not a bug.

Sure, okay, it can be that.  Again, I'll edit the title.
I tend to think of a bug in terms of something undocumented that goes against behavior a user might logically expect.  
Note, I think it _IS_ documented _NOW_, but it is a recent change to the user manual (per my discussions with the maintainer @pryrt).

**Comment by alankilborn** [comment]
I edited the title but now it feels awkward. :-(
I guess hard to express nicely with a short amount of text.

**Comment by Coises** [comment]
Looking at the code, either of these changes — using rectangular selections and/or using all selections in a multiple selection and not just the primary selection — would complicate the processing considerably.

There are oddities as it is. If you execute any of the trim spaces commands with a stream selection that doesn’t begin at the beginning of a line and/or doesn’t end at the end of a line, and anything is changed, the selection is also changed to begin at the beginning of the first line and end at the end of the last line. (If nothing is changed, the original stream selection is restored.) If you execute a trim spaces command with a multiple selection, only the primary selection remains after the command (and, if anything changed, the selection begins at the beginning of the first line and ends at the end of the last line).

The process is basically to save the (primary) selection; create a new selection that extends from the beginning of the first line to the end of the last line; apply a regular expression replacement; if nothing changed, restore the saved selection.

Saving a rectangular selection is possible (but different from saving a stream selection); replacing a rectangular selection with a stream selection from the beginning of the first line to the end of the last line isn’t much different from doing that with a stream selection; but if a change occurs... then what? Leave the stream selection? (Probably surprising to the user.) Convert back to a rectangular selection? — you can’t just restore it naïvely, because character positions will have changed; if it was a trim leading or trim both action, selecting the same *content* might not result in a selection that is rectangular anymore. (For trim trailing, it should be *possible* to reconstruct a rectangular selection encompassing the same content, taking into consideration that the anchor and/or caret might go into virtual space.)

For multiple selections, this would mean iterating through the selections (since regex replacements don’t work on multiple selections), noting the line numbers, trimming those lines and (somehow — possible, I think, but complicated) keeping track of the character position changes so the selection could be restored.

Just a personal opinion, but I don’t see the added complexity as being justified, unless there is a clear use case for this facility.

**Comment by alankilborn** [comment]
Of course I knew from the start it had a low likelihood of actually being implemented.
It would be _nice_ if it were, but...

**Comment by alankilborn** [comment]
Here's an interesting one:

Start with some text in a new tab:

```txt
one
two
three
four
five
six
seven
eight
nine
ten
```

Hold Alt while dragging the mouse to create a zero-width selection off to the right of some of the lines:

![image](https://github.com/user-attachments/assets/e503888e-06a2-4204-950a-680ed7a528c0)

Press space to get:

![image](https://github.com/user-attachments/assets/4d8df299-41e1-45b9-aea8-8697c4b9de49)

Execute a "trim trailing" command; note that nothing changes.

Press Backspace to remove one of the added spaces from each line touched by the selection.

Execute a "trim trailing" command again, to obtain:

![image](https://github.com/user-attachments/assets/f7ed9db0-eb0c-4c4a-9e67-7b1116b39a80)
